crates/proto/src/runtime.rs

```diff
@@ -352,6 +353,54 @@ pub trait Time {
             .unwrap()
             .as_secs()
     }
+
+    /// This implements a retry handler for jobs that might not complete successfully (e.g.,
+    /// DNS requests made via UDP.) It starts a worker future immediately, then every
+    /// retry_interval_time period up to a maximum of max_workers. It will immediately return
+    /// the first worker that completes successfully, or an error if no workers succeed.
+    /// It does not implement an overall timeout to bound the work -- use [`Self::timeout`] for
+    /// that.
+    async fn retry<F, Fut, T, E>(
+        worker: F,
+        retry_interval_time: Duration,
+        max_workers: usize,
+    ) -> Result<T, E>
+    where
+        F: Fn() -> Fut + Send,
+        Fut: Future<Output = Result<T, E>> + Send,
+        T: Send,
+        E: From<&'static str> + Send,
+    {
+        let mut futures = FuturesUnordered::new();
+
+        let retry_timer = Self::delay_for(retry_interval_time).fuse();
+        pin_mut!(retry_timer);
+
+        futures.push(worker());
+        let mut workers = 1;
+
+        loop {
+            futures_util::select! {
+                result = futures.next() => {
+                    match result {
+                        Some(result) => {
+                            return result;
+                        }
+                        None => {
+                            return Err(E::from("no workers successful"));
+                        }
+                    }
+                }
+                _ = &mut retry_timer => {
+                    if workers < max_workers {
+                        workers += 1;
+                        futures.push(worker());
+                        retry_timer.set(Self::delay_for(retry_interval_time).fuse());
+                    }
+                }
+            }
+        }
```

Comment:
Would this get simpler if we used something like tokio's [`Interval`](https://docs.rs/tokio/latest/tokio/time/struct.Interval.html)?